File: InsertionSort/c/insertion_sort.c


void insertion_sort(int* arr, int p, int r) {

  int i, j, t;

  for (i=p+1; i<r; i++) {
    j = i;
    t = arr[i];

    while ((j>p) && (arr[j] < arr[j-1])) {
      arr[j]=arr[j-1];
      j--;
    }

    arr[j] = t;
  }

}

File: InsertionSort/scala/InsertionSort.scala
package com.jgdodson.algorithms

object InsertionSort {

  def insertionSort(arr: Array[Int]): Unit = {

    insertionSort(arr, 0, arr.length)
  }

  def insertionSort(arr: Array[Int], p: Int, r: Int): Unit = {

    var j = 0
    var t = 0

    for (i <- p+1 until r) {
      t = arr(i)
      j = i-1

      while (j >= p && t < arr(j)) {
        arr(j+1) = arr(j)
        j -= 1
      }

      arr(j+1) = t
    }
  }

}
